src/modules/endpoint/sagas/fetchCategories.js

```diff
@@ -1,99 +1,79 @@
-    yield fork(fetchCategories, city, prioritised, urls)
-  }
-
-  const otherCodes = codes.filter(value => value !== prioritised)
+function * fetchLanguages (city: string): Saga<Array<LanguageModel>> {
+  const params = {city}
 
-  for (const code: string of otherCodes) {
-    yield fork(fetchCategories, city, code, urls)
-  }
-
-  return urls
+  const languagesPayload = yield call(() => request(languagesEndpoint, params))
+  return languagesPayload.data
 }
 
-function * fetchLanguageCodes (city: string): Saga<Array<string>> {
+function * fetchLanguagesAndCategories (database: MemoryDatabase, action: FetchCategoriesRequestActionType): Saga<void> {
+  const city: string = action.params.city
+
   try {
-    const params = {city}
+    const language: string = action.params.language
+    const urls: ResourceCacheType = {}
 
-    const languagesPayload = yield call(() => request(languagesEndpoint, params))
-    const languageModels: Array<LanguageModel> = languagesEndpoint.mapResponse(languagesPayload.data, params)
-    const codes = languageModels.map(model => model.code)
-    const success: LanguagesFetchSucceededActionType = {
-      type: 'LANGUAGES_FETCH_SUCCEEDED', payload: languagesPayload, city
-    }
-    yield put(success)
-    return codes
-  } catch (e) {
-    const failed: LanguagesFetchFailedActionType = {type: `LANGUAGES_FETCH_FAILED`, city, message: e.message}
-    yield put(failed)
-    throw e
-  }
-}
+    const start = performanceNow()
+    const languages = yield call(fetchLanguages, city)
+    const categoryModel = yield call(fetchCategoriesByLanguage, city, language)
+    const end = performanceNow()
+    console.warn(`fetch categories: ${end - start}ms`)
 
-function * fetch (action: FetchCategoriesRequestActionType): Saga<void> {
-  const city: string = action.params.city
+    findResources(categoryModel.toArray()).forEach(url => {
+      const hash = fnv.hash(url).hex()
+      urls[url] = `${OFFLINE_CACHE_PATH}/${city}/${hash}.${getExtension(url)}`
+    })
 
-  try {
-    const prioritisedLanguage: string = action.params.prioritisedLanguage
+    database.changeContext(new DataContext(city, language), categoryModel, languages, urls)
+
+    yield call(fetchResourceCache, city, urls)
 
-    const codes = yield call(fetchLanguageCodes, city)
-    const urls = yield call(fetchAllCategories, city, codes, prioritisedLanguage)
+    yield call(persistCategories, database)
 
     const success: CategoriesFetchSucceededActionType = {type: `CATEGORIES_FETCH_SUCCEEDED`, city}
     yield put(success)
-    yield call(downloadResources, city, Array.from(urls))
   } catch (e) {
     const failed: CategoriesFetchFailedActionType = {type: `CATEGORIES_FETCH_FAILED`, city, message: e.message}
     yield put(failed)
   }
 }
 
-export default function * fetchSaga (): Saga<void> {
-  yield takeLatest(`FETCH_CATEGORIES_REQUEST`, fetch)
+export default function * fetchCategories (database: MemoryDatabase): Saga<void> {
```

Comment:
Should also be renamed since it also fetches the languages.

Reply:
This will likely change a bit. Not really sure where and when we will fetch the languages. Probably with the categories. They will be stored somewhere in the state then.